### DISCUSSION TITLE
Not seeing the "Generate" widget in SysID tool under Beta 3 or 4

### DISCUSSION THREAD

**Discussion by mjhealy** [wpilibsuite/2024Beta#50] [Q&A]
I'm trying to put some sample code together for the kids on my team before the season starts, and I'm at the point where I need to profile the performance of one of our old drive bases and either one of our old Romi units or an XRP unit I purchased for evaluation.  (The [latest docs on SysID](https://docs.wpilib.org/en/latest/docs/software/pathplanning/system-identification/configuring-project.html) don't suggest that XRP support is available yet, but I'm hopeful.)

Unfortunately, when I open the SysID tool, I'm not seeing the "Generate" widget that would be used to specify the hardware configuration for the device to be profiled (which is obviously something of a blocker, to say the least).  It also doesn't appear in the list under the "Widgets" menu: only the Logger, Analyzer, and Program Log are shown there, though I'm also seeing the "Diagnostic Plots" widget within the SysID app window when I run it.

Resetting the workspace doesn't resolve the problem, nor does reinstalling the software (which I did this morning, when I found out that Beta4 was now available).

Is there something else that I can do to trigger the display of the Generator widget?  Or is this likely to be a bug, in which case I'll instead ask if there's something that I can do to provide additional information to help diagnose it?

Thanks!

**Comment by Kevin-OConnor** [comment]
The generation portion of SysID has been removed. It didn't work very well for most folks last year. There's a PR here to add some logging routines to WPILib that users can utilize, hopefully that will be ready to merge by Kickoff: https://github.com/wpilibsuite/allwpilib/pull/6033

We also need to get a piece in place to load data logged in the new format into SysID. I think we're still hopeful to get that done before or shortly after Kickoff as well. You can see some discussion of someone playing around with the proposed implementation on CD near the end of this thread: https://www.chiefdelphi.com/t/ok-dont-panic-sysid-is-half-going-away/443094/81

We recognize the frustration of this work being done at the last minute, but continuing to support the project generator became infeasible. It didn't work at all for some use cases like swerve drive motor characterization. It didn't work well for any use cases (there were glitches with it crashing or failing to connect to NT that folks couldn't manage to track down). We're hopeful that we can get something in place for this season that is less frustrating for teams to use.

**Comment by mjhealy** [comment]
OK, I can certainly see that.  I know that we'd run into problems getting the "generation" portion to work in the past in any given year, and it would take us some time to remember just how we'd had to do things the last time.  Getting the logging stuff moved into the WPILib, and then having the SysID tool (or some future offspring) just handle the data analysis seems like a reasonable step to take.

That said, since the point of the generator was to support exactly what was needed to perform the various data captures for analysis, I assume that there's likely to be some interface requirement for the logging routines to work with an arbitrary drive base (e.g., "your drive base needs to implement this interface in Java, or this concept/template-compatible interface in C++, in order to expose a function that provides direct control of the motors by means of specifying voltages for left and right sides", or something conceptually equivalent).  And the discussion in the CD thread (thanks for the link!) certainly seems to point in that direction.

Is there a design doc someplace that I can read through in order to see what the rough plan is and what I'd need to do to use whatever version of the routines is in Beta4, or at least some sort of hand-waving in the direction of a programmer's guide to using them?  (Given infinite time, I'm certainly capable of going code-diving to try to figure it out, but between EOY headaches at my day job, trying to get umpteen things ready for the kids on the team before kickoff, and -- oh, yeah -- the social/family obligations that come with the season, that's not a readily viable option.)

FWIW, I'm happy to be a test subject and provide early feedback, if it'll help!

**Reply by Oblarg** [reply]
There's no interface requirement, but you have to pass some callbacks.  An example of the team-side code is given in the thread.

It should be easier to write the callbacks than it was to configure the project generator, all told.

**Comment by mjhealy** [comment]
Good to know, though the <code>SysIdRoutine</code> interface that's referenced in the example in the thread doesn't appear to be included in the latest published version of the [beta WPILib docs](https://github.wpilib.org/allwpilib/docs/beta/java/index.html), which would seem to line up with [the pull request](https://github.com/wpilibsuite/allwpilib/pull/6033) that appears to define the interface not yet having been merged into the master branch.

I've read through the code in the PR, and it does look reasonably straightforward (at least in Java), though the devil will be in the details.

Is there something a little more detailed than the PR or the snippet from the CD thread, even in draft form, towards which I can point my student software lead?  (The team uses C++ for historical reasons, so that's obviously where their focus will be, and while they've got some limited exposure to Java, they're not comfortable doing the translations between languages, so even a high-level walkthrough as in some of the control system documentation for other elements of WPILib could be helpful.)

And I know that this is likely to be an unwanted question, but given the 11th hour nature of the change, I **have** to ask: is there any thought when this will be available in a beta for me to give it a try (even if it's just in Java), or is it more likely to be, "With kickoff nearly upon us, plus the holidays, we're still not sure if any version of this will be included in a cut of WPILib until after kickoff, even in an undocumented format"?

**Reply by PeterJohnson** [reply]
We are not planning on another beta before the kickoff release (probably released a few days before kickoff).  The PR is waiting on C++ tests before it is merged, but we are planning on having it in the kickoff release.

Right now the PR is all there is—docs have not been written yet.

**Reply by mjhealy** [reply]
OK.  I figured this was likely to be the case, but this _is_ "the season of hope".  (*chuckle*)

I'll keep my eyes peeled for the docs when they land, and will definitely be trying to put this to the test once it does.

Thanks!

**Comment by mjhealy** [comment]
The `SysIdRoutine` infrastructure made it into the 1st full release post-kickoff, which is great to see.  And we've been able to wire stuff into place to run the commands as needed, and the information appears to be getting captured in the WPILog (also great news).

Unfortunately the docs on how to carry that data over to the SysId tool still appear to be missing from the Control System documentation, which still describes [using the project generator](https://docs.wpilib.org/en/latest/docs/software/pathplanning/system-identification/configuring-project.html), rather than SysId.

It looks like the work on [getting support for log parsing built into SysId](https://www.chiefdelphi.com/t/ok-dont-panic-sysid-is-half-going-away/443094/90) was completed to some extent, in that I can select a WPILog file that I know (via `fgrep`) includes the tags for my drive base's results, but the SysId tool continues to say that it's in "theoretical analysis mode" and tells me to select a data file to analyze recorded data even after I've done that (and the tool shows some of the tags in the "Log Loader" window).

I know that y'all are working hard to get everything available for the teams, and there will presumably be more coming in a future update, but I'm hoping that there's some additional guidance (or at least a bug or a push request) that we can follow.

Thanks!

**Reply by sciencewhiz** [reply]
You can see some draft documentation here. https://github.com/wpilibsuite/frc-docs/pull/2492

**Reply by mjhealy** [reply]
Groovy: that's helpful, in that I can now get the "data selector" window populated with the test state, velocity, position, and voltage.  However, the "Analysis Type" menu only includes "Arm", "Elevator", and "Simple" options, none of which (of course) seem to fit for a drive base.  And the pull request still includes discussions of an option for profiling "Drivetrains" in source/docs/software/advanced-controls/system-identification/introduction.rst.

Is that still coming, or are we now going to be calculating the values for kS, etc. for each of the two sides separately?  (Which we can obviously do if we need to, but it would be good to know.)
